### ISSUE TITLE
2-players lightgun games

### ISSUE THREAD

**Issue by phasermaniac83** [libretro/opera-libretro#180]
Hi! Can this be possible?
There are some 3do 2-player games but especially Shootout at old Tucson would be amazing to play with a friend.
Thanks

**Comment by trapexit** [comment]
It's been some time since I worked on the lightgun. There is a bug in the implementation that needs to be worked on but I had thought that 2 player was working. I'll have to investigate. Do make sure that the controller player count is greater than 1 in the settings.

**Comment by phasermaniac83** [comment]
Many thanks for the info Trapexit.
I tested a little bit more:
Mad Dog 2 works for 2 gun players (3do Lightguns). But for me, Shootout old Tucson (Arcade Lightgun) player2 (left side) doesn't work. No aiming nor button presses in service menu or ingame.

**Comment by Darknior** [comment]
Cool thanks a lot for this report, i will try too. I'm also working for now on Gun configuration for all my emulators :)

**Comment by trapexit** [comment]
No. Not currently. IIRC it's a 5 pixel crosshair. Something to just get a sense of where you were. I just mimicked some other core when I wrote it. It is more intended for those mouse based lightguns where you wouldn't be using them. This is using a normal mouse? For games without mouse support?

**Comment by jdeath** [comment]
Thanks. Sorry to hijack, I'll look around in lr_input_crosshair.* and see if I can make it bigger than 5 pixels. I don't mind compiling my own. I am using it on an arcade setup with a mayflash/wii lightgun. Just starting to use this core, so only tried a couple games and not all the lightgun options. For the couple games I tried, the game crosshair only updates when I hit the trigger, thus the need for bigger crosshair.  The wii lightgun is not accurate enough to shoot without it.

**Comment by trapexit** [comment]
I will. I've been  busy with other 3DO development. The 3DO dev community is very small and I maintain a large portion of the projects. Only have so many cycles... but it's on my todo.

**Comment by Darknior** [comment]
> I will. I've been busy with other 3DO development. The 3DO dev community is very small and I maintain a large portion of the projects. Only have so many cycles... but it's on my todo.

Thanks a lot.
And please, if you can, add a real cross to the 3DO emulator. For now it is only a pixel and for my part i can't see it most of time and games are impossible to play :(

**Comment by phasermaniac83** [comment]
Many thanks @trapexit.
If you have lightguns @Darknior, I recommend you to play without crosshairs, this way is how (for me) lightgun games are really fun, because with a crosshair you actually don't aim, you only drag the cursor.

**Comment by Darknior** [comment]
> Many thanks @trapexit. If you have lightguns @Darknior, I recommend you to play without crosshairs, this way is how (for me) lightgun games are really fun, because with a crosshair you actually don't aim, you only drag the cursor.

Yes i know, but some time it help if the gun is not fine configured.
And i also play gun games with my little girl, and it is very hard for a little children to play without a cross. It help a lot to listen how to use a gun.

**Comment by trapexit** [comment]
> Thanks a lot.
> And please, if you can, add a real cross to the 3DO emulator. For now it is only a pixel and for my part i can't see it most of time and games are impossible to play :(

Do you have a core that does what you're talking about so I can use it as reference?

**Comment by jdeath** [comment]
lr-mame, lr-mame2003, nestopia, fceumm have it. 
Actually, most of the cores listed here: https://wiki.batocera.org/emulators:lightgun_games
except stella, vice64, mesen, bsnes
Alot of cores have option to turn off crosshair as well.

**Comment by Darknior** [comment]
> > Thanks a lot.
> > And please, if you can, add a real cross to the 3DO emulator. For now it is only a pixel and for my part i can't see it most of time and games are impossible to play :(
> 
> Do you have a core that does what you're talking about so I can use it as reference?

Nestopia can be a good exemple for you
https://github.com/libretro/nestopia/search?q=crosshair

**Comment by Tovarichtch** [comment]
The best examples for crosshairs are SNES9X, Nestopia and Genesis Plus GX. They are just perfect with the correct color and visibility.

**Comment by trapexit** [comment]
> Nestopia can be a good exemple for you https://github.com/libretro/nestopia/search?q=crosshair

https://github.com/libretro/nestopia/blob/9fc815e9d0cf6a166aa2e98c693cd90147d3c4c8/libretro/libretro.cpp#L221

They do pretty much what I do just with black and white and likely bigger.

**Comment by trapexit** [comment]
Sure, optional is fine. I didn't make it optional because it was added before mouse based guns became more popular and didn't need them. I'm just trying to get a preferred style. Provide similarity with another core.

**Comment by StormedBubbles** [comment]
Hello, I wish I could help more with this issue, but I did notice something in [this YouTube video](https://www.youtube.com/watch?v=qgWIA2ZPQYM&t=310s). The user is controlling player 1 (left) using the mouse/gun assigned to port 1 (blue crosshair). For me with lr-opera, the port 1 gun controls player 2 (right) in SAOT, and I am unable to control player 1 at all (just like someone else mentioned above).

The most recent changes to the lightgun code were the reload adjustments I made last year, but those should not have affected this behavior at all. Just throwing this out there in case it gives a hint to a solution for 2-player SAOT.

**Comment by Codes4Fun** [comment]
I discovered something interesting... currently opera-libretro writes 4 bytes, while portfolio_os reads 5 bytes !

https://github.com/trapexit/portfolio_os/blob/bee7b0c8e4287083c73cb5497b658880c9e7af8e/src/input/LightGunRom.c#L142

I tried modifying opera-libretro to 5 bytes with a zero, and on trigger the screen would flash but no hits... so I set it to one instead and it works as usually. I suspect that means the lightgun hit a scanline on the screen.

I suspect though this is why there may be issues with having a lightgun and a second lightgun/controller at the same time.

**Comment by trapexit** [comment]
Yes, I've mentioned somewhere that the version of the code implementing the lightgun in Opera is wrong. It was RE'ed in place before I got a signal analyzer and got OS code. I document all the details https://3dodev.com/documentation/hardware/opera/pbus

**Comment by StormedBubbles** [comment]
@Codes4Fun Hey, I actually got all 2-player lightgun features working for every NTSC console and arcade game (didn't test PAL) and also fixed the offscreen-reloading behavior in Mad Dog McCree and Space Pirates. I made core options to manually set offsets so that you don't have to calibrate. I figured my changes were not what was desired upstream, and I don't really have a full understanding of *why* it works, so I didn't try submitting as a pull request. But if this makes everything playable from what I can tell. I haven't seen anything breaking happen.

I called it a "hacks" branch because I figured that's exactly what I made 😆 

https://github.com/StormedBubbles/opera-libretro/tree/hacks

**Comment by Codes4Fun** [comment]
@StormedBubbles That's awesome ! I do see a potential overflow bug, the buffer size check wasn't updated to 5 for both functions:

https://github.com/StormedBubbles/opera-libretro/blob/55cad4d052080b0f3ef756e742a9baa941871cf7/libopera/opera_pbus.c#L156

I'll try it out !

**Comment by drewjbx1** [comment]
@StormedBubbles I came across this trying to find a solution for 2 player light gun support. I am mostly an end-user and know very little about coding, just simple configs and scripting. I did download your opera-hack but do not know how to compile it? I was hoping to simply download a modified opera_libretro.dll file for Win10. Any suggestions or tips would be appreciated!

**Comment by StormedBubbles** [comment]
Hey, you can compile this using the same instructions as for MAME. Those are the clearest instructions I could find out there on the web. I compiled it for Windows 11 64-bit and tested the result on my Windows 10 64-bit laptop. It seems to operate fine. I can't seem to post a dll here natively, so here's a link to download what I compiled:

https://pixeldrain.com/u/Hf3CCHfN

This contains the various lightgun edits I made for my own build on top of the latest upstream source from the main branch here. I want to reiterate that I do not know how to properly fix the lightgun system and that I happened to find through trial-and-error something that *appears* to work for me. My hope is that someone nabs the good parts of these edits and does a more accurate take on emulating the lightgun functionality of the 3DO and its derivative arcade hardware.

I think VLDP Pixel Format = XRG8888 and having 2 input devices active are the only changes to default core options I made on my system (I use a Raspberry Pi 5). Other than that, you should just have to enable the 3DO Lightgun (Arcade Lightgun for Shootout at Old Tucson and Mad Dog 2: The Lost Gold [arcade version]) for each player, have the appropriate BIOS, enable the "raw" input driver if using Windows, set your mouse indices, and have at it. For my own convenience, I added manual offset adjustments to the core options so that I wouldn't have to calibrate every game every time.

**Comment by drewjbx1** [comment]
Ok I will try this out, originally I do have the 1 lightgun working as the 2nd player with the official core. Would be nice to get this merged into the official for all the lightgun users out there as we always try to get the most possible gun games working correctly. Thanks for the good info and modified core!
